src/bootstrap/src/core/build_steps/test.rs

```diff
@@ -354,50 +356,129 @@ impl Step for RustAnalyzer {
     const DEFAULT: bool = true;
 
     fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
-        run.path("src/tools/rust-analyzer")
+        run.path("src/tools/rust-analyzer").alias("rust-analyzer")
     }
 
     fn make_run(run: RunConfig<'_>) {
         run.builder.ensure(Self { stage: run.builder.top_stage, host: run.target });
     }
 
-    /// Runs `cargo test` for rust-analyzer
+    #[cfg_attr(
+        feature = "tracing",
+        instrument(
+            level = "debug",
+            name = "RustAnalyzer::run",
+            skip_all,
+            fields(stage = self.stage, host = ?self.host),
+        ),
+    )]
+    fn run(self, builder: &Builder<'_>) {
+        // FIXME: until the whole RA test suite runs on `i686`, we only run `proc-macro-srv` tests.
+        if self.host.starts_with("i686") {
+            debug!(host = ?self.host, "skipping rust-analyzer main tests due to test suite not working on `i686`");
+            return;
+        }
+
+        debug!(stage = self.stage, host = ?self.host, "ensuring compiler");
+        let compiler = builder.compiler(self.stage, self.host);
+
+        debug!(stage = self.stage, host = ?self.host, "ensuring std");
+        builder.ensure(compile::Rustc::new(compiler, self.host));
+
+        let mut cargo = tool::prepare_tool_cargo(
+            builder,
+            compiler,
+            Mode::ToolRustc,
+            self.host,
+            Kind::Test,
+            "src/tools/rust-analyzer",
+            SourceType::InTree,
+            &["in-rust-tree".to_owned()],
+        );
+        cargo.allow_features(tool::RustAnalyzer::ALLOW_FEATURES);
+
+        cargo.env("CARGO_WORKSPACE_DIR", builder.src.join("src/tools/rust-analyzer"));
+
+        // RA's test suite tries to write to the source directory, that can't work in Rust CI.
+        cargo.env("SKIP_SLOW_TESTS", "1");
```

Comment:
Note that this will mean not all r-a tests would be run via `./x test rust-analyzer`

Reply:
Is there some blessed directory where writes are allowed? (I believe we are trying to write to temp dirs)

Reply:
Yes, bootstrap has a `builder.out` (which is basically some kind of `build/` directory). Under this, the structure could be something like (I only know exactly the compiletest-managed output dir shape, I don't quite remember the crate unit test output dir shape).

```
build/
    test/
        rust-analyzer/...
```

TL;DR: sure, if the writes are done specific against such build output dir (for exmaple, run-make tests are allowed to do writes to such per-run-make-test output dir).

Reply:
In r-l/r CI, the build output dir doesn't overlap with source checkout dir, where the source dir is read-only.

Reply:
Gotcha, sounds like something that ought to be easily supported but can also be done as a follow up (given it needs touch up on the r-a side)